include/boost/math/special_functions/detail/interval_prime_sieve.hpp

```diff
@@ -0,0 +1,337 @@
+    decltype(auto) NewRange(const Integer left, const Integer right, OutputIterator resultant_primes) noexcept;
+};
+
+template<typename Integer, typename OutputIterator>
+void IntervalSieve<Integer, OutputIterator>::Settdlimit() noexcept
+{
+    const double dr {static_cast<double>(right_)};
+    const double delta {static_cast<double>(delta_)};
+    const double tdest {delta * std::log(dr)};
+
+    // Small cases
+    if(tdest * tdest >= dr)
+    {
+        tdlimit_ = static_cast<std::size_t>(std::sqrt(dr));
+        plimit_ = 0;
+        return;
+    }
+
+    // First guess
+    if(tdest <= 1ul<<30)
+    {
+        tdlimit_ = static_cast<std::size_t>(tdest);
+    }
+
+    else
+    {
+        tdlimit_ = 1ul<<30;
+    }
+
+    // Find the corresponding prime
+    std::size_t i;
+    for(i = pss_.len - 1; i > 0; --i)
+    {
+        if(static_cast<double>(pss_.ps[i]) * tdlimit_ < dr)
+        {
+            break;
+        }
+    }
+    plimit_ = pss_.prime[i];
+
+    double tdlimit_guess = 1 + std::fmod(dr, static_cast<double>(pss_.ps[i]));
+    if(tdlimit_guess * tdlimit_guess >= dr)
+    {
+        tdlimit_ = static_cast<std::size_t>(std::sqrt(dr));
+        plimit_ = 0;
+    }
+}
+
+template<typename Integer, typename OutputIterator>
+void IntervalSieve<Integer, OutputIterator>::SeiveLength(const Integer d) noexcept
+{
+    Integer r {left_ % d};
+    Integer start {0};
+
+    if(r != 0)
+    {
+        start = d - r;
+    }
+
+    for(Integer i {start}; i >= 0 && i < b_.size(); i += d)
+    {
+        b_[static_cast<std::size_t>(i)] = 0;
+    }
+}
+
+template<typename Integer, typename OutputIterator>
+void IntervalSieve<Integer, OutputIterator>::Sieve() noexcept
+{
+    std::int_fast64_t primes_range {};
+    if(plimit_ <= 10)
+    {
+        primes_range = 10;
+    }
+
+    else
+    {
+        primes_range = plimit_;
+    }
+
+    // Sieve with pre-computed (or small) primes and then use the wheel for the remainder    
+    std::size_t i {};
+    Integer j;
+    if(plimit_ <= pss_.prime.back())
+    {
+        SeiveLength(static_cast<Integer>(2));
+        for(; pss_.prime[i] < primes_range; ++i)
+        {
+            SeiveLength(pss_.prime[i]);
+        }
+
+        j = w_.Next(pss_.prime[--i]);
+    }
+    
+    else
+    {
+        primes_.resize(static_cast<std::size_t>(prime_approximation(right_)));
+        linear_sieve(primes_range, primes_.begin());
+
+        for(; primes_[i] < primes_range; ++i)
+        {
+            SeiveLength(primes_[i]);
+        }
+
+        j = w_.Next(primes_[--i]);
+    }
+
+    for(; j <= tdlimit_; j = w_.Next(j))
+    {
+        SeiveLength(j);
+    }
+}
+
+template<typename Integer, typename OutputIterator>
+decltype(auto) IntervalSieve<Integer, OutputIterator>::WriteOutput() noexcept
+{
+    for(std::size_t i {}; i < b_.size(); ++i)
+    {
+        if(b_[i])
+        {
+            *resultant_primes_++ = left_ + i;
+        }
+    }
+    return resultant_primes_;
+}
```

Comment:
Hmmm, I can see, you're trying to maintain a consistent interface, right? But I think it's pushing you into an awkward design corner.
I would treat the output of `IntervalSieve` in the same way as Stepanov's `sift` function: the bitset `b_` _is_ the output. Don't process it any further: it has _all_ the information. The user can get prime number values from it with a trivial constant-time operation. Remember, the user provides the value of `left_` in the first place.

I suspect, although I could be wrong, that the vector `primes_` could be removed completely if the linear sieve wrote directly to the bitset? Ah, I just realized that the container version in another file takes these low primes in pre-calculated? Not sure why there is a design difference there, but doesn't matter too much.

I think you can cut through the whole output iterator vs container argument by just making the bitset be the output.
I'm not sure if that holds for the more advanced algorithms, I'll look at them next.

Reply:
You would still need the vector `primes_`. The intervals only represent a range that can be fit into L1 not the entire range from the user. In order to do this you need to recursively pass a reference to the computed primes which is exactly what you see in the container version. In the OI version you don't have the capability to do that so it maintains the vector `primes_` of only exactly what it needs to account for that shortfall. It also resets the bitset every range so the bitset has to get written to a buffer anyway; seems reasonable to me to give the actual underlying value instead of a boolean.

Reply:
So, with `primes_`, as I said in another comment, it's not strictly a container, it's just an iterator range, so don't conflate it with the output iterator vs output container issue. It's only when `emplace_back()` and `resize()`, etc appear that something really has to be a container. So, I hope we can remove that difference between the two `IntervalSieve`s.

Well, if the bitset is the output, that removes the last difference between the two `IntervalSieve`s, right? Because all the output iterator/container stuff gets removed. So that's a win for code simplicity.

The other reason is also about general design simplicity, and in particular I mean the S in SOLID:  single responsibility.
What does `IntervalSieve` do currently? It sieves primes in a bitset AND then transforms that bitset into a sequence of primes. That's two responsibilities: g(f(x)) (or g ∘ f), where g is the transform. Yes, at the moment you 'know' that you want the prime number sequence from the sieve, but, I guess part of the point of separating the two operations is that it gives you the freedom to change your mind. You might discover that in one case you need g(y), but in another it's actually h(y). Anyway, that's enough on that. I think my first reason is more important.

Reply:
The two `IntervalSieve`s need not be reconciled. There are currently two complete, and totally discrete implementations of the entire prime sieve with different interfaces: one with the dual interface and one strictly OI. You are now suggesting a different interface where the user receives boolean values. There is no function in the sieve that currently does that so once again all the methods and handling would need retooled. I am uninterested in making any more design or interface changes without clear consensus and direction. I do not know if this a question for the 4 reviewers (you, @NAThompson, @pabristow, and @jzmaddock), or if it needs to be taken to the mailing list to be resolved. This entire PR has been, and will continue to be stagnant until the interface question is resolved.